src/Native/Runtime/windows/CoffNativeCodeManager.cpp

```diff
@@ -0,0 +1,499 @@
+    // flags from mrt. Call to this method is used as a cue to mark the method info
+    // as execution aborted. Note - if pMethodInfo was cached, this scheme would not work.
+    //
+    // If the method has EH, then JIT will make sure the method is fully interruptible
+    // and we will have GC-info available at the faulting address as well.
+
+    CoffNativeMethodInfo * pNativeMethodInfo = (CoffNativeMethodInfo *)pMethodInfo;
+    pNativeMethodInfo->executionAborted = true;
+
+    return;
+}
+
+struct CoffEHEnumState
+{
+    PTR_UInt8 pEHInfo;
+    UInt32 uClause;
+    UInt32 nClauses;
+};
+
+// Ensure that CoffEHEnumState fits into the space reserved by EHEnumState
+static_assert(sizeof(CoffEHEnumState) <= sizeof(EHEnumState), "CoffEHEnumState too big");
+
+bool CoffNativeCodeManager::EHEnumInit(MethodInfo * pMethodInfo, PTR_VOID * pMethodStartAddress, EHEnumState * pEHEnumStateOut)
+{
+    assert(pMethodInfo != NULL);
+    assert(pMethodStartAddress != NULL);
+    assert(pEHEnumStateOut != NULL);
+
+    CoffNativeMethodInfo * pNativeMethodInfo = (CoffNativeMethodInfo *)pMethodInfo;
+    CoffEHEnumState * pEnumState = (CoffEHEnumState *)pEHEnumStateOut;
+
+    size_t unwindDataBlobSize;
+    PTR_VOID pUnwindDataBlob = GetUnwindDataBlob(m_moduleBase, pNativeMethodInfo->mainRuntimeFunction, &unwindDataBlobSize);
+
+    uint8_t unwindBlockFlags = *(dac_cast<DPTR(uint8_t)>(pUnwindDataBlob) + unwindDataBlobSize);
+
+    // return if there is no EH info associated with this method
+    if ((unwindBlockFlags & UBF_FUNC_HAS_EHINFO) == 0)
+    {
+        return false;
+    }
+
+    *pMethodStartAddress = dac_cast<PTR_VOID>(m_moduleBase + pNativeMethodInfo->mainRuntimeFunction->BeginAddress);
+
+    pEnumState->pEHInfo = dac_cast<PTR_UInt8>(pUnwindDataBlob) + unwindDataBlobSize + 1;
+    pEnumState->uClause = 0;
+    pEnumState->nClauses = VarInt::ReadUnsigned(pEnumState->pEHInfo);
+
+    return true;
+}
+
+bool CoffNativeCodeManager::EHEnumNext(EHEnumState * pEHEnumState, EHClause * pEHClauseOut)
+{
+    assert(pEHEnumState != NULL);
+    assert(pEHClauseOut != NULL);
+
+    CoffEHEnumState * pEnumState = (CoffEHEnumState *)pEHEnumState;
+    if (pEnumState->uClause >= pEnumState->nClauses)
+        return false;
+    pEnumState->uClause++;
+
+    pEHClauseOut->m_tryStartOffset = VarInt::ReadUnsigned(pEnumState->pEHInfo);
+
+    UInt32 tryEndDeltaAndClauseKind = VarInt::ReadUnsigned(pEnumState->pEHInfo);
+    pEHClauseOut->m_clauseKind = (EHClauseKind)(tryEndDeltaAndClauseKind & 0x3);
+    pEHClauseOut->m_tryEndOffset = pEHClauseOut->m_tryStartOffset + (tryEndDeltaAndClauseKind >> 2);
+
+    // For each clause, we have up to 4 integers:
+    //      1)  try start offset
+    //      2)  (try length << 2) | clauseKind
+    //      3)  if (typed || fault || filter)    { handler start offset }
+    //      4a) if (typed)                       { index into type table }
```

Comment:
It would be nice to use a type table index scheme where the type table is per-object file and the code manager can find it from a header.  Perhaps a work item to this effect would be warranted?